src/app/pcasts/dao/notifications_dao.py

```diff
@@ -0,0 +1,55 @@
+import json
+from . import *
+from app.pcasts.dao import series_dao, episodes_dao
+
+def new_episode_notifications(user_id, max_episodes, offset):
+  subscriptions = get_subscribed_series_for_new_episodes(user_id)
+  series_ids = [s.series_id for s in subscriptions]
+  episodes = episodes_dao.get_episodes_by_many_series(user_id, series_ids,  \
+      offset, max_episodes)
+  # Populate series data
+  series = series_dao.get_multiple_series(series_ids, user_id)
+  series_map = {s.id: s for s in series}
+  for e in episodes:
+    e.series = series_map[e.series_id]
+    e.unread_notifcation = is_unread_notification(user_id, e.id)
+  return episodes
+
+def is_unread_notification(user_id, episode_id):
+  read = ReadNewEpisodeNotification.query. \
+      filter(ReadNewEpisodeNotification.user_id == user_id,\
+             ReadNewEpisodeNotification.episode_id == episode_id).first()
+  return read is None
+
+def get_subscribed_series_for_new_episodes(user_id):
+  return Subscription.query.filter(Subscription.user_id == user_id and \
+      Subscription.subscribed_new_episodes == True).all()
+
+def update_new_episode_has_read(user_id, episode_ids):
+  series_ids = episodes_dao.get_series_ids_from_episodes(episode_ids)
+  subscriptions = subscriptions_dao.get_user_subscriptions_for_series(user_id,\
+      series_ids)
+
+  series_to_subscription = {}
+  for subscription in subscriptions:
+    series_to_subscription[subscription.series_id] = subscription.id
+
+  episode_to_subscription = {}
+  for episode_id in episode_ids:
+    episode = Episode.query.filter(Episode.id == episode_id).first()
```

Comment:
why do we do this if we already know the episode_id?

Reply:
We start out with just episode ids which are provided by the user.
Then we get the subscriptions for those episode id_s which are unorered.
Now we need to map from episode_id to subscription so that when we create a ReadNewEpisodeNotification object we can provide the subscription_id.
The only way to do this is to establish a relationship between episode id is to establish from id to series and then from series to subscription and to chain that mapping. In this case episode_to_subscription is the final mapping which creates the first mapping on the fly since we do no have the actual episode objects but only the ids